sdk/python/kfp/components/v1_modelbase.py

```diff
@@ -0,0 +1,394 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+from collections import OrderedDict
+from collections import abc
+from typing import (Any, Dict, List, Mapping, MutableMapping, MutableSequence,
+                    Sequence, Type, TypeVar, Union, cast, get_type_hints)
+
+T = TypeVar('T')
+
+
+def verify_object_against_type(x: Any, typ: Type[T]) -> T:
+    """Verifies that the object is compatible to the specified type (types from
+    the typing package can be used)."""
+    #TODO: Merge with parse_object_from_struct_based_on_type which has almost the same code
+    if typ is type(None):
+        if x is None:
+            return x
+        else:
+            raise TypeError('Error: Object "{}" is not None.'.format(x))
+
+    if typ is Any or type(typ) is TypeVar:
+        return x
+
+    try:  #isinstance can fail for generics
+        if isinstance(x, typ):
+            return cast(typ, x)
+    except Exception:
+        pass
+
+    if hasattr(typ, '__origin__'):  #Handling generic types
+        if typ.__origin__ is Union:  #Optional == Union
+            exception_map = {}
+            possible_types = typ.__args__
+            if type(
+                    None
+            ) in possible_types and x is None:  #Shortcut for Optional[] tests. Can be removed, but the exceptions will be more noisy.
+                return x
+            for possible_type in possible_types:
+                try:
+                    verify_object_against_type(x, possible_type)
+                    return x
+                except Exception as ex:
+                    exception_map[possible_type] = ex
+                    pass
+            #exception_lines = ['Exception for type {}: {}.'.format(t, e) for t, e in exception_map.items()]
+            exception_lines = [str(e) for t, e in exception_map.items()]
+            exception_lines.append(
+                'Error: Object "{}" is incompatible with type "{}".'.format(
+                    x, typ))
+            raise TypeError('\n'.join(exception_lines))
+
+        #not Union => not None
+        if x is None:
+            raise TypeError(
+                'Error: None object is incompatible with type {}'.format(typ))
+
+        #assert isinstance(x, typ.__origin__)
+        generic_type = typ.__origin__ or getattr(
+            typ, '__extra__', None
+        )  #In python <3.7 typing.List.__origin__ == None; Python 3.7 has working __origin__, but no __extra__  TODO: Remove the __extra__ once we move to Python 3.7
+        if generic_type in [
```

Comment:
nit: We have dropped Python 3.6 support. So we should be able to clean this up.
Feel free to do it here or in a follow-up PR.

Reply:
I agree.

I will do this in a second PR because of the complexity of the code here, that way if any bugs are introduced they belong to a different commit.